### PR TITLE
OHRM5X-2064:implement graph api

### PR DESCRIPTION
**Pull request by ShehanOHRM**

### CODE REVIEW

File: src/screens/time/AttendancePickEmployee.tsx [hunk 1]
```diff
@@ -43,13 +43,18 @@ import {
   pickSubordinate,
 } from 'store/time/attendance/actions';
 import {selectSubordinates} from 'store/time/attendance/selectors';
+import {debounce, DebouncedFunc} from 'lodash';
 
 class AttendancePickEmployee extends React.Component<
   AttendancePickEmployeeProps,
   AttendancePickEmployeeState
 > {
+  fetchSubordinates: DebouncedFunc<(text: string) => {}>;
```
Comment (RajithaKumara) [inline src/screens/time/AttendancePickEmployee.tsx:52]:
```suggestion
  fetchSubordinates: DebouncedFunc<(text: string) => {}>;

```
add new line

File: src/lib/helpers/attendance.ts [hunk 2]
```diff
@@ -598,73 +604,88 @@ const getDatesStringKey = (name: string) => {
...
-      arr1.push(obj);
+      leaveArray.push(leaveObject);
+    }
+  });
+
+  return leaveArray;
+};
+
+const getGraphObject = (
+  graphData: GraphRecordsDetailsArray[],
```
Comment (RajithaKumara) [inline src/lib/helpers/attendance.ts]:
```suggestion
  graphData: GraphRecordsDetails[],
```

File: src/lib/helpers/attendance.ts [hunk 3]
```diff
@@ -598,73 +604,88 @@ const getDatesStringKey = (name: string) => {
...
+  workWeekData: WorkSummaryGraphObject[],
+  leaveResponse: GraphRecordsLeaveObject[],
+) => {
+  let totalWorkDuration = 0;
+  let totalLeaveHours = 0;
+
+  if (graphData) {
+    const graphDataArray: GraphRecordsDetailsObject[] = [];
```
Comment (RajithaKumara) [inline src/lib/helpers/attendance.ts]:
```suggestion
    const graphDataArray: GraphRecordsDetails[] = [];
```

File: src/screens/time/AttendancePickEmployee.tsx [hunk 4]
```diff
@@ -43,13 +43,18 @@ import {
...
   constructor(props: AttendancePickEmployeeProps) {
     super(props);
+    this.fetchSubordinates = debounce((text: string) => {
```
Comment (RajithaKumara) [inline src/screens/time/AttendancePickEmployee.tsx:56]: :ok_hand:

File: src/screens/time/AttendanceSummary.tsx [hunk 5]
```diff
@@ -72,6 +72,7 @@ import {
   EmployeeAttendanceSummaryScreenRouteParams,
 } from 'screens/time/navigators';
 import {getFirstAndLastNames} from 'lib/helpers/name';
+import {selectMyInfo} from '../../store/auth/selectors';
```
Comment (RajithaKumara) [inline src/screens/time/AttendanceSummary.tsx]:
```suggestion
import {selectMyInfo} from 'store/auth/selectors';
```

File: src/screens/time/AttendanceSummary.tsx [hunk 6]
```diff
@@ -117,7 +118,7 @@ class AttendanceSummary extends React.Component<
       startDayIndex: this.state.startDayIndex,
       employeeAttendance,
       employeeName: this.props.route.params?.employeeAttendance
-        ? getFirstAndLastNames(this.props.route.params.employeeAttendance)
+        ? getFirstAndLastNames(this.props.route.params?.employeeAttendance)
```
Comment (RajithaKumara) [inline src/screens/time/AttendanceSummary.tsx]:
```suggestion
        ? getFirstAndLastNames(this.props.route.params.employeeAttendance)
```

File: src/services/endpoints.ts [hunk 7]
```diff
@@ -31,6 +31,8 @@ export const API_ENDPOINT_ATTENDANCE_GRAPH =
   '/api/v2/leave/employees/leave-requests';
 export const API_ENDPOINT_ATTENDANCE_WORK_SUMMARY =
   '/api/v2/dashboard/employees/time-at-work';
+export const API_ENDPOINT_GRAPH_LEAVE_RECORDS =
```
Comment (RajithaKumara) [inline src/services/endpoints.ts]:
```suggestion
export const API_ENDPOINT_EMPLOYEE_LEAVE =
```
:thinking:

File: src/store/time/attendance/sagas.ts [hunk 8]
```diff
@@ -132,14 +135,12 @@ function* fetchLeaveRecords(action: FetchLeaveRecordsAction) {
...
-        API_ENDPOINT_LEAVE,
+        API_ENDPOINT_GRAPH_LEAVE_RECORDS,
         {},
         {
           fromDate: action.payload.fromDate,
           toDate: action.payload.toDate,
-          pendingApproval: true,
-          scheduled: true,
-          taken: true,
+          statuses: ['1', '2', '3'],
```
Comment (RajithaKumara) [inline src/store/time/attendance/sagas.ts]:
```suggestion
          statuses: ['1', '2', '3'], // pending approval, scheduled, taken
```
will add comment like this

File: src/store/time/attendance/sagas.ts [hunk 9]
```diff
@@ -179,41 +180,62 @@ function* fetchAttendanceGraphRecords(
...
+        prepare(
+          API_ENDPOINT_ATTENDANCE_WORK_SUMMARY,
+          {},
+          {
+            timezoneOffset: getCurrentTimeZoneOffset(),
+            currentDate: action.payload.fromDate,
+            currentTime: '00:00',
+            ...(action.payload.empNumber && {
+              empNumber: action.payload.empNumber,
+            }),
```
Comment (RajithaKumara) [inline src/store/time/attendance/sagas.ts]: :thinking:

File: src/store/time/attendance/sagas.ts [hunk 10]
```diff
@@ -179,41 +180,62 @@ function* fetchAttendanceGraphRecords(
...
+            includeEmployees: 'onlyCurrent',
+            statuses: ['1', '2', '3'],
+            ...(action.payload.empNumber && {
+              empNumber: action.payload.empNumber,
+            }),
+          },
+        ),
+      );
+
+    const workWeekResponse: ApiResponse<WorkSummaryGraphObject[]> =
```
Comment (RajithaKumara) [inline src/store/time/attendance/sagas.ts]: I'm not clear on this

File: src/store/time/attendance/types.ts [hunk 11]
```diff
@@ -264,6 +304,8 @@ export interface SingleEmployeeAttendance {
   empNumber: number;
   firstName: string;
   lastName: string;
+
```
Comment (RajithaKumara) [inline src/store/time/attendance/types.ts]:
```suggestion
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: ios/Carthage/Checkouts/AppAuth-iOS/.codecov.yml [hunk 12]
```diff
@@ -0,0 +1,2 @@
+ignore:
```
Comment (RajithaKumara) [inline ios/Carthage/Checkouts/AppAuth-iOS/.codecov.yml:1]: any idea about this?

File: src/lib/helpers/attendance.ts [hunk 13]
```diff
@@ -340,8 +344,8 @@ const dayMapper = {
   saturday: 'Sa',
 };
 
-const getLeaveColourById = (id: string) => {
-  return LEAVE_TYPE_COLORS[parseInt(id, 10) % LEAVE_TYPE_COLORS.length];
+const getLeaveColourById = (id: number) => {
+  return LEAVE_TYPE_COLORS[parseInt(String(id), 10) % LEAVE_TYPE_COLORS.length];
```
Comment (RajithaKumara) [inline src/lib/helpers/attendance.ts]:
```suggestion
  return LEAVE_TYPE_COLORS[id % LEAVE_TYPE_COLORS.length];
```
:thinking:

File: src/lib/helpers/attendance.ts [hunk 14]
```diff
@@ -598,73 +602,76 @@ const getDatesStringKey = (name: string) => {
...
+};
+
+const getGraphObject = (
+  workData: WorkSummaryGraphObject[],
+  leaves: GraphRecordsLeaveObject[],
+  totalWorkDuration: TotalTimeObject,
+) => {
+  let totalLeaveHours = 0;
+
+  if (leaves) {
```
Comment (RajithaKumara) [inline src/lib/helpers/attendance.ts]:
```suggestion
```
I don't think we need this condition

File: src/lib/helpers/attendance.ts [hunk 15]
```diff
@@ -598,73 +602,76 @@ const getDatesStringKey = (name: string) => {
...
-          leave: [],
-        },
-        [getDatesStringKey(workWeekData[6].workDay.day)]: {
-          workHours:
-            workWeekData[6].totalTime.hours +
-            workWeekData[6].totalTime.minutes / 60,
-          leave: [],
-        },
-      };
+    if (workData) {
```
Comment (RajithaKumara) [inline src/lib/helpers/attendance.ts]:
```suggestion
```
:thinking:

File: src/screens/time/AttendanceDetails.tsx [hunk 16]
```diff
@@ -62,6 +62,7 @@ import {
   convertDateObjectToStringFormat,
   getWeekDayFromIndex,
 } from 'lib/helpers/attendance';
+import {selectMyInfo} from '../../store/auth/selectors';
```
Comment (RajithaKumara) [inline src/screens/time/AttendanceDetails.tsx]:
```suggestion
import {selectMyInfo} from 'store/auth/selectors';
```

File: src/screens/time/AttendancePickEmployee.tsx [hunk 17]
```diff
@@ -68,7 +73,7 @@ class AttendancePickEmployee extends React.Component<
   }
 
   onRefresh = () => {
-    this.props.fetchSubordinates();
+    this.fetchSubordinates('');
   };
```
Comment (RajithaKumara) [inline src/screens/time/AttendancePickEmployee.tsx:78]: *

File: src/store/time/attendance/sagas.ts [hunk 18]
```diff
@@ -179,40 +184,45 @@ function* fetchAttendanceGraphRecords(
   try {
     yield openLoader();
 
-    const graphResponse: ApiResponse<GraphRecordsObject> = yield apiCall(
+    const workWeekResponse: ApiResponse<
```
Comment (RajithaKumara) [inline src/store/time/attendance/sagas.ts]:
```suggestion
    const workSummaryResponse: ApiResponse<
```

File: src/store/time/attendance/types.ts [hunk 19]
```diff
@@ -18,11 +18,12 @@
  *
  */
 
-import {LeaveStatus} from 'store/leave/leave-usage/types';
 import {WorkWeek, Holiday} from 'store/leave/common-screens/types';
 import {Moment} from 'moment';
 
 export const DEFAULT_START_DAY = 1;
+export const ADMIN = 'Admin';
+export const ESS = 'ESS';
```
Comment (RajithaKumara) [inline src/store/time/attendance/types.ts]: do we still need this

File: src/store/time/attendance/types.ts [hunk 20]
```diff
@@ -148,13 +151,63 @@ export interface WorkSummaryObject {
...
+  leaveType: LeaveTypeGraph;
+  lengthHours: number;
+  date: string;
+}
+
+export interface TotalWorkDuration {
+  currentWeek: currentWeekObject;
+}
+
+export interface currentWeekObject {
```
Comment (RajithaKumara) [inline src/store/time/attendance/types.ts]:
```suggestion
export interface CurrentWeekObject {
```
refactor rename

File: src/store/time/attendance/types.ts [hunk 21]
```diff
@@ -148,13 +151,63 @@ export interface WorkSummaryObject {
...
+  totalTime: {
+    hours: number;
+    minutes: number;
+  };
+}
+
+export interface GraphRecordsDetailsArray {
+  leaveType: LeaveTypeGraph;
+  noOfDays: number;
+}
```
Comment (RajithaKumara) [inline src/store/time/attendance/types.ts]: do we need this
```suggestion
```

File: src/store/time/attendance/types.ts [hunk 22]
```diff
@@ -148,13 +151,63 @@ export interface WorkSummaryObject {
...
+
+export interface GraphRecordsDetailsObject {
```
Comment (RajithaKumara) [inline src/store/time/attendance/types.ts]:
```suggestion
export interface GraphRecordsDetails {
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: ios/Carthage/Checkouts/AppAuth-iOS/.codecov.yml (hunk 12 above)
Comment (ShehanOHRM) [reply ios/Carthage/Checkouts/AppAuth-iOS/.codecov.yml:1]: This file using a IOS auth login flow

**Pull request merged**